src/libstd/sys/unix/thread.rs

```diff
@@ -125,16 +125,32 @@ impl Thread {
     }
 
     pub fn sleep(dur: Duration) {
+        fn get_secs(secs: &mut u64) -> libc::time_t {
+            if *secs > libc::time_t::max_value() as u64 {
+                *secs -= libc::time_t::max_value() as u64;
+                libc::time_t::max_value()
+            } else {
+                let copy = *secs;
+                *secs = 0;
+                copy as libc::time_t
+            }
+        }
+
+        let mut secs = dur.as_secs();
         let mut ts = libc::timespec {
-            tv_sec: dur.as_secs() as libc::time_t,
+            tv_sec: 0,
             tv_nsec: dur.subsec_nanos() as libc::c_long,
         };
 
         // If we're awoken with a signal then the return value will be -1 and
         // nanosleep will fill in `ts` with the remaining time.
         unsafe {
-            while libc::nanosleep(&ts, &mut ts) == -1 {
-                assert_eq!(os::errno(), libc::EINTR);
+            while secs > 0 {
+                ts.tv_sec = get_secs(&mut secs);
+                if libc::nanosleep(&ts, &mut ts) == -1 {
```

Comment:
I think that this'll need to be a loop of some form to correctly handle EINTR. The `nanosleep` function will automatically update `ts` with the remaining time when `EINTR` is returned, and I think on success as well. Basically we'll need to repeatedly call `nanosleep` until it succeeds.